libdnf/dnf-context.cpp

```diff
@@ -2271,16 +2271,30 @@ dnf_context_setup(DnfContext *context,
 
     /* setup a file monitor on the rpmdb, if we're operating on the native / */
     if (g_strcmp0(priv->install_root, "/") == 0) {
-        rpmdb_path = g_build_filename(priv->install_root, "var/lib/rpm/Packages", NULL);
-        file_rpmdb = g_file_new_for_path(rpmdb_path);
-        priv->monitor_rpmdb = g_file_monitor_file(file_rpmdb,
-                               G_FILE_MONITOR_NONE,
-                               NULL,
-                               error);
-        if (priv->monitor_rpmdb == NULL)
-            return FALSE;
-        g_signal_connect(priv->monitor_rpmdb, "changed",
-                         G_CALLBACK(dnf_context_rpmdb_changed_cb), context);
+        rpmdb_path = rpmGetPath("%{_dbpath}", NULL);
+        if (rpmdb_path && !g_file_test(rpmdb_path, G_FILE_TEST_EXISTS))
+                g_clear_pointer(&rpmdb_path, free);
```

Comment:
Does this work without directly putting a monitor on the rpmdb file? We're putting monitors on files normally, so we should probably also query rpm for what kind of rpmdb it is and build the path for the database file to monitor.

Reply:
That's probably a question for @ffesti or @pmatilai, but I know you can ask rpm for what type of database it is, and the filenames are fixed for databases per rpmdb type.

Reply:
> the filenames are fixed for databases per rpmdb type.

Even if they are fixed, the backends can change in time, which requires changes here, like the bug about the path. The get() for the path makes it generic. I left the other code paths just in case, but I doubt they will be used with any recent rpm.

I looked around and did not find anything direct for the file name. I see there are these: `Packages`, `rpmdb.sqlite` and `Packages.db` for various backends (the dummydb backend uses `NULL`).

The `rpmdbFStat()` uses `db->db_ops->path` in combination with `rpmGenPath(rpmdbHome(db), dbfile, NULL);` to get the file for the `stat` call, but these internals (`db->db_ops`) seem to be hidden to the library user.

Reply:
Ping reviewers

Reply:
I think the way to fix this is to interrogate RPM for the db type and track the right file, and fallback to not establishing a monitor at all.

Reply:
Okay, I rewrote it. It's with no fallback, as you suggested. The things depend on the reliability of the RPM macros and the file names or backend names did not change on the RPM side - sadly, there is no public API to get this low level information - or at least I did not find anything for it.

Reply:
you should be able to just read `%_db_backend` and `%_db_path` variables to get the database backend and the path the files will be written to. The names for the files is hard-coded to the backend, so you can map based on the database backend being used.

Reply:
yes, that's what I do and what the "rewrite" is about. Still the hardcoded thing in the RPM, not in the libdnf. This "libdnf expect certain names and knows of certain RPM db backends" thing breaks abstraction. Not much can be done, when there is no API on the RPM side to be used (or I do not know of any such at least).